### PR TITLE
BinaryFlow and AntiJoin nodes, hookup DSL not

### PR DESCRIPTION
**Pull request by ibdknox**
This change adds an AntiJoin node type that serves as the basis for blocks that contain a logical negation in them. It also hooks up the DSLNot node that we created and adds a simple test to show what using not looks like.

### CODE REVIEW

File: src/runtime/dsl.ts [hunk 1]
```diff
@@ -531,8 +530,28 @@ class DSLBlock {
...
     // @TODO: Once we start having aggregates, we'll need to do some stratification here
     // instead of just throwing everything into a single JoinNode.
-    nodes.unshift(new runtime.JoinNode(constraints))
+    let join:Node = new runtime.JoinNode(constraints);
+
+    // @NOTE: We need to unify all of our sub-blocks along with ourselves
+    //        before the root node can allocate registers.
+    for(let not of this.nots) {
+      not.compile(items);
+      let notJoinNode = not.block.nodes[0];
```
Comment (joshuafcole) [inline src/runtime/dsl.ts]: We should note that this isn't sufficient in the post-aggregate world too.

File: src/runtime/runtime.ts [hunk 2]
```diff
@@ -1243,6 +1243,103 @@ export class InsertNode implements Node {
...
+  abstract onLeft(index:Index, prefix:ID[], transaction:number, round:number, results:Iterator<ID[]>):void;
+  abstract onRight(index:Index, prefix:ID[], transaction:number, round:number, results:Iterator<ID[]>):void;
+}
+
+export class AntiJoin extends BinaryFlow {
+  leftIndex = new IntermediateIndex();
+  rightIndex = new IntermediateIndex();
+  keyFunc:KeyFunction;
+
+  constructor(public left:Node, public right:Node, public keyRegisters:Register[]) {
```
Comment (joshuafcole) [inline src/runtime/runtime.ts]: What happens when not contains an agg here?

File: src/runtime/dsl.ts [hunk 3]
```diff
@@ -531,8 +530,28 @@ class DSLBlock {
     }
     // @TODO: Once we start having aggregates, we'll need to do some stratification here
     // instead of just throwing everything into a single JoinNode.
-    nodes.unshift(new runtime.JoinNode(constraints))
+    let join:Node = new runtime.JoinNode(constraints);
+
+    // @NOTE: We need to unify all of our sub-blocks along with ourselves
+    //        before the root node can allocate registers.
+    for(let not of this.nots) {
+      not.compile(items);
```
Comment (joshuafcole) [inline src/runtime/dsl.ts]: We should address the way nots embed their parents logic here in a comment

Comment (ibdknox) [inline src/runtime/dsl.ts]: 👍

File: src/runtime/dsl.ts (hunk 1 above)
Comment (ibdknox) [inline src/runtime/dsl.ts]: 👍

File: src/runtime/runtime.ts (hunk 2 above)
Comment (ibdknox) [inline src/runtime/runtime.ts]: I'm going to add a LinearFlow node that allows us to stack these guys up arbitrarily.

**Pull request merged**